### PR TITLE
Add option to use react-windowed-select

### PR DESCRIPTION
**Pull request by peterjcoffman**
Hoist P/R Checklist
-------------------

Review and check off the below. Items that do not apply can also be checked off to indicate they
have been considered. If unclear if a step is relevant, please leave unchecked and note in comments.

- [x] Up to date with `develop` branch as of last change.
- [x] Ready for review (or open as a draft PR / add `wip` label).
- [x] Added CHANGELOG entry (or N/A)
- [x] Reviewed for breaking changes (add `breaking-change` label + CHANGELOG if so, or N/A)
- [x] Updated doc comments / prop-types (or N/A)
- [x] Reviewed and tested on Mobile (or N/A)
- [x] Created Toolbox branch / PR (link provided here, or N/A)
https://github.com/xh/toolbox/pull/317

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by peterjcoffman** [comment]
To improve performance of Select on large lists, per this issue:
https://github.com/xh/hoist-react/issues/1458

**Comment by lbwexler** [comment]
@amcclain -- any concerns here?  I would like to merge this and have us all start kicking the tires.  As @haynesjm42 can attest, this is a pretty serious bug

File: kit/react-select/index.js [hunk 1]
```diff
@@ -5,10 +5,10 @@
  * Copyright © 2019 Extremely Heavy Industries Inc.
  */
 import {elemFactory} from '@xh/hoist/core';
-import Select from 'react-select';
-import AsyncSelect from 'react-select/async';
-import AsyncCreatable from 'react-select/async-creatable';
-import Creatable from 'react-select/creatable';
+import Select from 'react-windowed-select';
+import AsyncSelect from 'react-windowed-select';
```
Comment (lbwexler) [inline kit/react-select/index.js]: This can all be on one line now

File: package.json [hunk 2]
```diff
@@ -63,7 +63,8 @@
     "regenerator-runtime": "~0.13.3",
     "router5": "~7.0.2",
     "router5-plugin-browser": "~7.0.2",
-    "store2": "~2.10.0"
+    "store2": "~2.10.0",
+    "react-windowed-select": "^2.0.1"
```
Comment (amcclain) [inline package.json]: It looks like we should *not* be specifying `react-select` as a direct dependency ourselves - i.e. we should remove it from this list. This new library specifies it as a dep (and notably not a peer-dep) with a particular version.  We only want one copy of `react-select` in our project and it should be the version this new library will work with.

Also, we want to use the `~` semver qualifier for our deps - this allows auto-upgrades of patch versions only, not minor versions.  (We have seen that minor versions are more likely to break, despite what the authors might claim with their version number choice.)

Finally, these deps should remain in alpha-order, and any time we change package.json we need to run `yarn` and check-in the modifications to `yarn.lock` as part of the change.

File: kit/react-select/index.js (hunk 1 above)
Comment (amcclain) [reply kit/react-select/index.js]: Don't we need to export `WindowedSelect` and use that in our `Select` component for us to get any advantage from this change???

As per their readme they re-export all the exports from react-select unmodified, but then also export their custom component. Apologies if I am missing something obvious here 😄

**Comment by amcclain** [comment]
I would also ask that you please:
* Add a Toolbox example with a large number of generated options (1000?  What kind of item count do we think will still be reasonable in terms of performance?)
* Test the mobile implementation (uses the same library)
* Add a note in the Select component header re. the use of `WindowedSelect`
* Add a CHANGELOG entry 

Those will let us tick off a few more of our PR checklist items.  Thanks 🙏

**Comment by peterjcoffman** [comment]
In the previous state, this was broken for everything except 'normal' selects- create and async did not work at all. Now they should work, I believe.

**Comment by peterjcoffman** [comment]
>     * Add a Toolbox example with a large number of generated options (1000?  What kind of item count do we think will still be reasonable in terms of performance?)

The branch https://github.com/xh/toolbox/tree/bigSelect has a new select with 5000 items on app/forms/inputs.

**Comment by peterjcoffman** [comment]
I am not sure menuIsOpen works correctly, i.e. the onKeyDown logic at Select:240-Select:249 does not seem to work properly.

**Comment by peterjcoffman** [comment]
I have found out in more detail what is wrong with `menuIsOpen`.

react-windowed-select does not export a creatable or async `WindowedSelect`. It exports only a 'basic' `WindowedSelect`. Therefore, I have to make Creatable and Async `WindowedSelect` myself.

I do this here:

```

const Select = WindowedSelect;
const AsyncSelect = (makeAsyncSelect(Select));
const AsyncCreatable = (makeCreatableSelect(AsyncSelect));
const Creatable = (makeCreatableSelect(Select));
```

`makeAsyncSelect` and `makeCreatableSelect` are functions exported from react-select which add creatable or async functionality to an existing select.

The problem is that there is an additional step in creating an async or creatable Select. Take a look at the very bottom of this file:
https://github.com/JedWatson/react-select/blob/master/packages/react-select/src/Creatable.js

```
const SelectCreatable = makeCreatableSelect<ElementConfig<typeof Select>>(
  Select
);

export default manageState<ElementConfig<typeof SelectCreatable>>(
  SelectCreatable
);
```

`SelectCreatable` is wrapped in a [StateManager](https://react-select.com/props#statemanager-props), which is what handles the [controllable props](https://github.com/JedWatson/react-select#controllable-props). We do not have access to the `manageState` function, so we cannot wrap our Selects in a `StateManager` and cannot get its functionality.

This breaks at least the functionality of `menuIsOpen`. On `Select.js`, the following branch (lines 243-247) will never be taken on a creatable or async Select:
```
    const {menuIsOpen} = this.reactSelectRef.current ? this.reactSelectRef.current.state : {};
    if (menuIsOpen && (e.key == 'Escape' || e.key == 'Enter')) {
        console.log('menu is open');
        e.stopPropagation();
    }
```
It may also break the functionality of `onChange`, `onInputChange`, and [the other things listed as controllable props](https://github.com/JedWatson/react-select#controllable-props). I believe this won't be a problem, since we already handle `onChange`, `value`, `onInputChange`, and `inputValue` ourselves.

**Comment by peterjcoffman** [comment]
Something else noteworthy: the library's handling of the StateManager is also buggy! Compare the code for Async and the code for Creatable:

https://github.com/JedWatson/react-select/blob/master/packages/react-select/src/Creatable.js
```
const SelectCreatable = makeCreatableSelect<ElementConfig<typeof Select>>(
  Select
);

export default manageState<ElementConfig<typeof SelectCreatable>>(
  SelectCreatable
);
```
It first makes the Select creatable, then wraps it in a StateManager.

https://github.com/JedWatson/react-select/blob/master/packages/react-select/src/Async.js
```
const SelectState = manageState<ElementConfig<typeof Select>>(Select);

export default makeAsyncSelect<ElementConfig<typeof SelectState>>(SelectState);
```
It first wraps the Select in a StateManager, then makes it async.

They do not do the same thing. And the implementation for Async is broken! Add a print statement inside the branch on line 243-247:
```
                const {menuIsOpen} = this.reactSelectRef.current ? this.reactSelectRef.current.state : {};
                if (menuIsOpen && (e.key == 'Escape' || e.key == 'Enter')) {
                    console.log('menu is open')
                    e.stopPropagation();
                }
```
The expected behavior is that hitting Escape while the Select is open will cause 'menu is open' to be printed to the console. This works as expected for a normal Select and a creatable Select, but not for an async Select.

**Comment by peterjcoffman** [comment]
Lee and I discussed this, and we believe it's appropriate to just swallow the keydown events even if the menu is closed. This means we no longer need to care about whether or not the menu is open, so my concerns are probably not a problem.

**Comment by peterjcoffman** [comment]
@amcclain 
>Test the mobile implementation (uses the same library)
I have tested the mobile implementation using the Selects available in the mobile version of Toolbox. However, I do not know how to test the Async or Creatable selects on mobile.

**Comment by lbwexler** [comment]
Peter, Colin, and I discussed.  There are issues with using this for the asyn/ createable/async creatable cases, and known bugs around monitoring the menu open state, that will effects how we handle key presses.

Shame to let this cause problems for the majority of simple selects by trying to force its use automatically.

Safer to expose this as an option -- ` useWindowedSelect` that can be opt-in for apps.  We should document it as having known limitations, and primarily for non-async non-creatable cases.

**Comment by lbwexler** [comment]
I think this is ok for merging now.  Its a very safe change now, that should have no impact, unless the options is set.  Toolbox example has some issues we could go over on monday

**Comment by lbwexler** [comment]
After discussing with Anselm late on Friday, we decided to go with a safer, and more minimal change that only effects basic selects.  We can always revisit, and remove this restriction, if needed and if react-windowed-select fixes its bugs.

Also tested large option lists on mobile.  Weirdly, performance for large option lists seems to be fine, so not adding the prop to the mobile component at this time.

**Pull request merged**